### PR TITLE
Band to tridiagonal local version.

### PR DESCRIPTION
**Pull request by rasolca**
Added local reduction band -> tridiagonalization.

See https://doi.org/10.1177/1094342013502097 for details about the algorithm.

- [x] local tridiagonalization
- [x] test
- [x] documentation

### CODE REVIEW

**Review by Sely85 — COMMENTED** [review]
A few minor comments.

**Review by albestro — APPROVED** [review]
LGTM.

I was not able to go through all the algorithm details. Just some comments in addition to what I already pointed out in the code:

- If I see correctly, more internal classes member functions can be tagged as `const noexcept` (e.g. `BandBlock`, `SweepWorker`);
- `workers` looks like an array of `Pipeline`s. It may be an overkill, it's just an idea I wanted to point out so you can think about it;
- At the moment I don't have any better idea than what you had about the lambdas to get access to the shared_pointer `a_ws` and keeping it alive as long as needed (for the async task).

File: include/dlaf/eigensolver/band_to_tridiag/mc.h [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+        blas::copy(n_d, (BaseType<T>*) a_ws->ptr(0, start), inc, tile_t.ptr({0, 0}), tile_t.ld());
+        blas::copy(n_e, (BaseType<T>*) a_ws->ptr(1, start), inc, tile_t.ptr({1, 0}), tile_t.ld());
+      };
+
+      const auto size = mat_trid.size().cols();
+      const auto nb = mat_trid.blockSize().cols();
+      if (sweep % nb == nb - 1 || sweep == size - 1) {
+        const auto tile_index = sweep / nb;
+        const auto start = tile_index * nb;
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/eigensolver/band_to_tridiag/mc.h:340]: It is an integer division, therefore sweep / nb * nb != sweep.

File: test/unit/eigensolver/test_band_to_tridiag.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+using namespace dlaf::test;
+using namespace testing;
+
+template <typename Type>
+class EigensolverBandToTridiagTest : public ::testing::Test {};
+
+TYPED_TEST_SUITE(EigensolverBandToTridiagTest, MatrixElementTypes);
+
+const std::vector<std::tuple<SizeType, SizeType, SizeType>> sizes = {
+    //{0, 2, 2},                              // m = 0
```
[the remark this answers is not in the document]
Comment (rasolca) [reply test/unit/eigensolver/test_band_to_tridiag.cpp:39]: Nope, still trying to understand why this case fails...

File: include/dlaf/eigensolver/band_to_tridiag/mc.h [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+template <class T>
+struct BandToTridiag<Backend::MC, Device::CPU, T> {
+  static SizeType nrSweeps(SizeType size) {
+    // Complex needs an extra sweep to have a real tridiagonal matrix.
+    return std::is_same<T, ComplexType<T>>::value ? size - 1 : size - 2;
+  }
+
+  static SizeType nrStepsForSweep(SizeType sweep, SizeType size, SizeType band) {
+    // Last Complex sweep should be handled differently.
+    return sweep == size - 2 ? 1 : util::ceilDiv(size - sweep - 2, band);
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/eigensolver/band_to_tridiag/mc.h:260]: Is the new comment better?

**Comment by rasolca** [comment]
@albestro Please note that I didn't change this PR to Pipeline because of #417.
The needed changes are instead included in branch https://github.com/eth-cscs/DLA-Future/tree/rasolca/tridiag_local_pipeline.

File: include/dlaf/eigensolver/band_to_tridiag.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+/// Real Type:
+/// HHT(0, 0) HHT(0, 1) .. HHT(0, last(0)) HHT(1, 0) .. HHT(1, last(1)) .. HHT(m - 3, 0).
+/// Complex Type:
+/// HHT(0, 0) HHT(0, 1) .. HHT(0, last(0)) HHT(1, 0) .. HHT(1, last(1)) .. HHT(m - 3, 0) HHT(m - 2, 0)
+/// where m is the size of A and HHT(sw, st) is the HH transformation given by (Id - v tau v^H),
+/// last(sw) = ceilDiv(m - sw - 2, band_size), (in case of complex types last(m - 2) = 1)
+/// and tau and v are defined in the following way:
+/// - start = 1 + sw + st * band_size
+/// - len = min(band_size, m - start)
+/// - pos = sw / band_size + st) * band_size
```
Comment (Sely85) [inline include/dlaf/eigensolver/band_to_tridiag.h:43]:
```suggestion
/// - pos = sw / (band_size + st) * band_size
```
?

**Review by Sely85 — APPROVED** [review]
LGTM

**Comment by rasolca** [comment]
I added the changes to use the pipeline (as #423 was merged).
Still not using Senders and receivers as HPX 1.8.0 as type erased senders are needed.

**Comment by rasolca** [comment]
bors try.

**Comment by rasolca** [comment]
bors try

**Comment by rasolca** [comment]
bors try

**Comment by rasolca** [comment]
bors try

File: include/dlaf/eigensolver/band_to_tridiag/mc.h [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+      else {
+        while (deps.size() < max_deps_size) {
+          deps.push_back(sf);
+        }
+      }
+    }
+
+    // Maximum size / (2b-1) sweeps can be executed in parallel.
+    const auto max_workers =
+        std::min(ceilDiv(size, 2 * b - 1), to_signed<SizeType>(get_num_threads("default")));
```
Comment (albestro) [inline include/dlaf/eigensolver/band_to_tridiag/mc.h:326]:
```suggestion
        std::min(ceilDiv(size, 2 * b - 1), to_SizeType(get_num_threads("default")));
```

File: include/dlaf/eigensolver/band_to_tridiag/mc.h [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+    vector<Pipeline<SweepWorker<T>>> workers;
+    workers.reserve(max_workers);
+    for (SizeType i = 0; i < max_workers; ++i)
+      workers.emplace_back(SweepWorker<T>(size, b));
```
Comment (albestro) [inline include/dlaf/eigensolver/band_to_tridiag/mc.h:330]:
```suggestion
    vector<Pipeline<SweepWorker<T>>> workers(max_workers, SweepWorker<T>(size, b));
```

**Review by albestro — APPROVED** [review]
Not knowing the details of the algorithm, I just gave a general look at the code. Mainly minor comments and suggestions (there are also various points where you can apply `const` to variables).

The main point of my comment is about where do we want to put the shared functions `nrSweeps` and `nrStepPerSweep`?

File: include/dlaf/eigensolver/band_to_tridiag/mc.h (hunk 6 above)
Comment (rasolca) [reply include/dlaf/eigensolver/band_to_tridiag/mc.h:330]: Pipeline<SweepWorker<T>> is not copyable, therefore the constructor you suggest cannot be used.

Comment (msimberg) [reply include/dlaf/eigensolver/band_to_tridiag/mc.h:330]: It should, however, probably be `emplace_back(size, b)` with the loop version?

Comment (albestro) [reply include/dlaf/eigensolver/band_to_tridiag/mc.h:330]: It is a `vector<Pipeline<SweepWorker>>`, which is something I missed at the beginning 😄, so I would say no.

Comment (msimberg) [reply include/dlaf/eigensolver/band_to_tridiag/mc.h:330]: Ah yes, so did I...!

**Review by teonnik — APPROVED** [review]
LGTM!

File: test/unit/eigensolver/test_band_to_tridiag.cpp [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  set(mat_local, [](auto) { return T{0}; });
+
+  for (SizeType j = 0; j < m - 1; ++j) {
+    mat_local({j, j}) = mat_trid_local({0, j});
+    mat_local({j + 1, j}) = mat_trid_local({1, j});
+    mat_local({j, j + 1}) = mat_trid_local({1, j});
+  }
+  mat_local({m - 1, m - 1}) = mat_trid_local({0, m - 1});
+
+  auto mat_v_local = matrix::test::allGather(lapack::MatrixType::General, mat_v);
```
[the remark this answers is not in the document]
Comment (rasolca) [reply test/unit/eigensolver/test_band_to_tridiag.cpp:74]: V has be be modifiable (see https://github.com/eth-cscs/DLA-Future/blob/dd869456d725dc6f51b718cd93d20d2e1d6edf1f/test/unit/eigensolver/test_band_to_tridiag.cpp#L77)

**Comment by rasolca** [comment]
bors try

**Pull request merged**